blog/2025/10/julia-1.12-highlights.md

```diff
@@ -0,0 +1,423 @@
+    return 0
+end
+
+end
+```
+
+```bash
+juliac --output-exe app_test_exe --bundle build --trim=safe --experimental ./AppProject
+```
+
+```bash
+./build/bin/app_test_exe
+Hello World!
+
+ls -lh build/bin/app_test_exe
+-rwxr-xr-x@ 1 gabrielbaraldi  staff   1.1M Oct  6 17:22 ./build/bin/app_test_exe*
+```
+
+## Redefinition of constants (structs)
+*Keno Fischer*, *Tim Holy*
+
+Bindings now participate in the "world age" mechanism previously used for methods. This has the effect that constants and structs can be properly redefined. As an example:
+
+```julia
+# Define a struct and a method on that struct:
+julia> struct Foo
+          a::Int
+       end
+
+julia> g(f::Foo) = f.a^2
+g (generic function with 1 method)
+
+julia> g(Foo(2))
+4
+
+# Redefine the struct (julia pre-1.12 would error here)
+julia> struct Foo
+          a::Int
+          b::Int
+       end
+
+# Note that functions need to be redefined to work on the new `Foo`
+julia> g(Foo(1,2))
+ERROR: MethodError: no method matching g(::Foo)
+The function `g` exists, but no method is defined for this combination of argument types.
+
+Closest candidates are:
+  g(::@world(Foo, 39296:39300)) # <- This is syntax for accessing the binding in an older "world"
+   @ Main REPL[2]:1
+
+julia> g(f::Foo) = f.a^2 + f.b^2
+g (generic function with 2 methods)
+
+julia> g(Foo(2,3))
+13
+```
+
+There is also work in progress in Revise.jl to automatically redefine functions on replaced bindings.
+This should significantly reduce the number of times you have to restart Julia while iterating on some piece of code.
+
+## New tracing flags and macros for inspecting what Julia compiles
+*Ian Butterworth*, *Nathan Daly*
+
+`--trace-compile-timing` is a new command-line flag that augments `--trace-compile` by printing how long each compiled method took (in milliseconds) before the corresponding `precompile(...)` line. This makes it easier to spot costly compilations.
+
+In addition, two macros for ad-hoc tracing without restarting Julia has been added:
+
+* `@trace_compile expr` runs `expr` with `--trace-compile=stderr --trace-compile-timing` enabled, emitting timed `precompile(...)` entries only for that call.
+* `@trace_dispatch expr` runs `expr` with `--trace-dispatch=stderr` enabled, reporting methods that are dynamically dispatched.
+
+**Examples**
+
+```julia
+julia> @trace_compile @eval rand(2,2) * rand(2,2)
+#=   79.9 ms =# precompile(Tuple{typeof(Base.rand), Int64, Int64})
+#=    4.4 ms =# precompile(Tuple{typeof(Base.:(*)), Array{Float64, 2}, Array{Float64, 2}})
+2×2 Matrix{Float64}:
+ 0.302276  0.14341
+ 0.738941  0.396414
+
+julia> f(x) = x
+
+julia> @trace_dispatch map(f, Any[1,2,3])
+precompile(Tuple{Type{Array{Int64, 1}}, UndefInitializer, Tuple{Int64}})
+precompile(Tuple{typeof(Base.collect_to_with_first!), Array{Int64, 1}, Int64, Base.Generator{Array{Any, 1}, typeof(Main.f)}, Int64})
+3-element Vector{Int64}:
+ 1
+ 2
+ 3
+
+```
+
+## New multi-threading features
```

Comment:
I added a new section to group all multi-threaded related changes, similar to the Pkg section. I'm not sure if the atomics one also belongs to here, I'm happy for the section title to be revised if someone comes up with a more appropriate name

Reply:
The title you chose seems good to me.